### PR TITLE
feat(vm): add support for memory `hugepages` and `keep_hugepages` attributes

### PR DESCRIPTION
**Pull request by sergelogvinov**
Allow use hugepages for VM memory allocation.

### Contributor's Note

- [x] I have added / updated documentation in `/docs` for any user-facing features or additions.
- [x] I have added / updated templates in `/example` for any new or updated resources / data sources.
- [x] I have ran `make example` to verify that the change works as expected.

### Proof of Work

```tf
resource "proxmox_virtual_environment_vm" "hugepages" {
  ...

  memory {
    dedicated = 188416
    hugepages = "1024"
  }
}
```

Memory page table on host machine:

before (without huge pages):

```shell
# cat /proc/meminfo | grep PageTables
PageTables:       872252 kB
SecPageTables:    853344 kB
```

after:

```shell
# cat /proc/meminfo | grep PageTables
PageTables:        14736 kB
SecPageTables:       120 kB
```

### Community Note

- Please vote on this pull request by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original pull request comment to help the community and maintainers prioritize this request
- Please do not leave "+1" or other comments that do not add relevant new information or questions, they generate extra noise for pull request followers and do not help prioritize the request

Relates [#1156](https://github.com/bpg/terraform-provider-proxmox/issues/1156)

### CODE REVIEW

File: proxmoxtf/resource/vm/vm.go [hunk 1]
```diff
@@ -4680,7 +4737,6 @@ func vmUpdate(ctx context.Context, d *schema.ResourceData, m interface{}) diag.D
 		updateBody.CPUCores = &cpuCores
 		updateBody.CPUSockets = &cpuSockets
 		updateBody.CPUUnits = &cpuUnits
-		updateBody.CPUAffinity = &cpuAffinity
```
Comment (bpg-dev) [inline proxmoxtf/resource/vm/vm.go:4765]: Was this accidentally removed, or is something wrong with updating the affinity?

File: docs/resources/virtual_environment_vm.md [hunk 2]
```diff
@@ -408,6 +408,14 @@ output "ubuntu_vm_public_key" {
...
         to `0`).
     - `shared` - (Optional) The shared memory in megabytes (defaults to `0`).
+    - `hugepages` - (Optional) Enable/disable hugepages memory (defaults to disable).
+        - `2` - 2MB hugepages.
+        - `1024` - 1GB hugepages.
+        - `any` - Any hugepages.
+    - `keep_hugepages` - (Optional) Keep hugepages memory after the VM is stopped (defaults
+        to `false`).
+
+    Settings `hugepages` and `keep_hugepages` are only allowed for `root@pam` authenticated user.
```
Comment (bpg-dev) [inline docs/resources/virtual_environment_vm.md:418]: It also requires `numa` to be enabled. We can add it to the doc an also to the schema, using `RequiredWith` field on `mkMemoryHugepages` entry.

File: proxmoxtf/resource/vm/vm.go [hunk 3]
```diff
@@ -1026,6 +1032,23 @@ func VM() *schema.Resource {
...
+					mkMemoryHugepages: {
+						Type:        schema.TypeString,
+						Description: "Enable/disable hugepages memory",
+						Optional:    true,
+						Default:     dvMemoryHugepages,
+						ValidateDiagFunc: validation.ToDiagFunc(validation.StringInSlice([]string{
+							"1024",
+							"2",
+							"any",
+						}, true)),
```
Comment (bpg-dev) [inline proxmoxtf/resource/vm/vm.go:1045]:
```go
RequiredWith: []string{"cpu.numa"}
```
(not tested)

**Review by bpg — COMMENTED** [review]
Hey @sergelogvinov! Thanks for another great feature!

I wasn't able to fully test it in my env though, probably just need more memory on my test pve box.
So I'd really appreciate a proof of work attached to the PR :) 

But I also noticed that resource update flow doesn't seem to be detecting errors on the apply. I.e. if I try to create a VM with `hugepages` enabled and `numa` disabled, it errors out correctly. But if I create a VM without `numa` first, then update the plan to add `hugepages`, the VM couldn't start, although the plan apply succeeds without showing any errors.
This might be nothing to do with your code, perhaps an old issue found its way up. But if you had a chance to debug it, that would be awesome!

I've also made few other comments to the PR, could you pls take a look?

Thanks!

File: proxmoxtf/resource/vm/vm.go (hunk 3 above)
Comment (sergelogvinov) [reply proxmoxtf/resource/vm/vm.go:1045]: Yep, i miss that, thanks

File: proxmoxtf/resource/vm/vm.go (hunk 1 above)
Comment (sergelogvinov) [reply proxmoxtf/resource/vm/vm.go:4765]: Probably we need to make fix-pr. It fix the previous PR, we set updateBody.CPUAffinity and next we trying to delete it  https://github.com/bpg/terraform-provider-proxmox/pull/1158/files#diff-32de2693719f6210316f28f1bde63075061724d6220adf157f96d3cb239a518cL4771

File: docs/resources/virtual_environment_vm.md (hunk 2 above)
Comment (sergelogvinov) [reply docs/resources/virtual_environment_vm.md:418]: Thanks, I added additional requirements in the documentation and code. 👍

**Comment by sergelogvinov** [comment]
> Hey @sergelogvinov! Thanks for another great feature!
> 
> I wasn't able to fully test it in my env though, probably just need more memory on my test pve box. So I'd really appreciate a proof of work attached to the PR :)
> 
> But I also noticed that resource update flow doesn't seem to be detecting errors on the apply. I.e. if I try to create a VM with `hugepages` enabled and `numa` disabled, it errors out correctly. But if I create a VM without `numa` first, then update the plan to add `hugepages`, the VM couldn't start, although the plan apply succeeds without showing any errors. This might be nothing to do with your code, perhaps an old issue found its way up. But if you had a chance to debug it, that would be awesome!
> 
> I've also made few other comments to the PR, could you pls take a look?
> 
> Thanks!

Better to reboot the server before tests. The kernel has to have huge free blocks of memory to allocate hugepages. Sometime you can get an error, because all memory was already defragmented.

The bast way to solve it - is allocate the hugepages on boot time. 
By command `echo 2 > /sys/devices/system/node/node3/hugepages/hugepages-1048576kB/nr_hugepages`
or kernel param `default_hugepagesz=1G hugepagesz=1G hugepages=2`

File: proxmoxtf/resource/vm/vm.go (hunk 1 above)
Comment (bpg-dev) [reply proxmoxtf/resource/vm/vm.go:4765]: ah, right, good catch!

**Review by bpg — APPROVED** [review]
LGTM! 🚀

**Pull request merged**